snapshot/snapshot_computation.py

```diff
@@ -0,0 +1,305 @@
+        else:
+            self._file_name = "output.hdf5"
+        self._output_file_path = os.path.join(
+            self._output_subdirectory, self._file_name
+        )
+        self._data_storage.create_file(self._output_file_path)
+        self._logger.info("Output file created: {}".format(self._output_file_path))
+        self._local_number_of_sims = len(self._macro_parameters)
+        self._logger.info(
+            "Number of local micro simulations = {}".format(self._local_number_of_sims)
+        )
+
+        if self._local_number_of_sims == 0:
+            if self._is_parallel:
+                self._logger.info(
+                    "Rank {} has no micro simulations and hence will not do any computation.".format(
+                        self._rank
+                    )
+                )
+                self._is_rank_empty = True
+            else:
+                raise Exception("Snapshot has no micro simulations.")
+
+        nms_all_ranks = np.zeros(self._size, dtype=np.int64)
+        # Gather number of micro simulations that each rank has, because this rank needs to know how many micro
+        # simulations have been created by previous ranks, so that it can set
+        # the correct global IDs
+        self._comm.Allgatherv(np.array(self._local_number_of_sims), nms_all_ranks)
+
+        # Get global number of micro simulations
+        self._global_number_of_sims = np.sum(nms_all_ranks)
+
+        # Create lists of local and global IDs
+        sim_id = np.sum(nms_all_ranks[: self._rank])
+        self._global_ids_of_local_sims = []  # DECLARATION
+        for i in range(self._local_number_of_sims):
+            self._global_ids_of_local_sims.append(sim_id)
+            sim_id += 1
+
+        self._micro_problem = getattr(
+            importlib.import_module(
+                self._config.get_micro_file_name(), "MicroSimulation"
+            ),
+            "MicroSimulation",
+        )
+
+        self._micro_sims_have_output = False
+        if hasattr(self._micro_problem, "output") and callable(
+            getattr(self._micro_problem, "output")
+        ):
+            self._micro_sims_have_output = True
+
+    def _solve_micro_simulations(self, micro_sims_input: dict) -> dict | None:
+        """
+        Solve all micro simulations and assemble the micro simulations outputs in a list of dicts format.
+
+        Parameters
+        ----------
+        micro_sims_input : dict
+            Dict in which keys are names of data and the values are the data which are required inputs to
+            solve a micro simulation.
+
+        Returns
+        -------
+        micro_sims_output : dict | None
+            Dicts in which keys are names of data and the values are the data of the output of the micro
+            simulations. The return type is None if the simulation has crashed.
+        """
+        # If time is a parameter, split input as time is a separate argument
```

Comment:
What is the best way to handle `dt`? If micro simulations require `dt` should it be a macro parameter, to allow it to change over different microsimulations or a fixed value specified in the config file?  What should `dt` be set to, if a microsimulation is independent of `dt`, e.g. the micro dummies?

Reply:
@IshaanDesai Shouldn't this also be a user input for "normal" coupled simulations?  To allow subcycling of the micro sims?

Reply:
In addition, I think not all micro simulations require a time step, i.e., have an explicit time dependence, especially mechanical problems (used for the snapshot computation). However, the Micro Manager currently requires a time step to be set, which, if we have a micro simulation in Python, we could just set to `None`  but a C++ implementation requires a float, which I don't think is optimal. The structure of the micro simulation template makes it necessary to set a time step, even if the micro simulation does not depend on it.

Reply:
Currently we do not allow the micro simulations to have their own time step, but force the macro time step on the micro simulations. Subcycling makes sense, and the time step needs to be an input parameter in the coupled simulations already.

Reply:
The micro time window size is being added as a parameter to the `coupling_params` group in https://github.com/precice/micro-manager/pull/112